deps/rabbit/src/rabbit_variable_queue.erl

```diff
@@ -1830,26 +2177,40 @@ publish1(Msg = #basic_message { is_persistent = IsPersistent, id = MsgId },
     UC1 = gb_sets_maybe_insert(NeedsConfirming, MsgId, UC),
     stats({1, 0}, {none, MsgStatus1}, 0,
           State2#vqstate{ next_seq_id = SeqId + 1,
+                          %% @todo Update next_deliver_seq_id if IsDelivered is true
+                          %% @todo Assert when doing the update to be sure there are no "deliver" holes.
+                          next_deliver_seq_id = case {IsDelivered, SeqId} of
+                            {true, NextDeliverSeqId} -> NextDeliverSeqId + 1;
+                            _ -> NextDeliverSeqId
+                          end,
                           in_counter  = InCount1,
                           unconfirmed = UC1 });
 publish1(Msg = #basic_message { is_persistent = IsPersistent, id = MsgId },
              MsgProps = #message_properties { needs_confirming = NeedsConfirming },
              IsDelivered, _ChPid, _Flow, PersistFun,
              State = #vqstate { mode                = lazy,
+                                version             = Version,
                                 qi_embed_msgs_below = IndexMaxSize,
                                 next_seq_id         = SeqId,
+                                next_deliver_seq_id = NextDeliverSeqId,
                                 in_counter          = InCount,
                                 durable             = IsDurable,
                                 unconfirmed         = UC,
                                 delta               = Delta}) ->
     IsPersistent1 = IsDurable andalso IsPersistent,
-    MsgStatus = msg_status(IsPersistent1, IsDelivered, SeqId, Msg, MsgProps, IndexMaxSize),
+    MsgStatus = msg_status(Version, IsPersistent1, IsDelivered, SeqId, Msg, MsgProps, IndexMaxSize),
     {MsgStatus1, State1} = PersistFun(true, true, MsgStatus, State),
     Delta1 = expand_delta(SeqId, Delta, IsPersistent),
     UC1 = gb_sets_maybe_insert(NeedsConfirming, MsgId, UC),
     stats(lazy_pub, {lazy, m(MsgStatus1)}, 1,
           State1#vqstate{ delta       = Delta1,
                           next_seq_id = SeqId + 1,
+                          %% @todo Update next_deliver_seq_id if IsDelivered is true
+                          %% @todo Assert when doing the update to be sure there are no "deliver" holes.
```

Comment:
I'm not sure what I meant by this. Spend some time to figure this out, this sounds important.